### DISCUSSION TITLE
Distinguishing role change on partial oldMember in guildMemberUpdate

### DISCUSSION THREAD

**Discussion by pranavbadami** [discordjs/discord.js#8409] [General]
### Which package is this bug report for?

discord.js

### Issue description

Hi folks! 

My bot's use case is that when a role is added to a member, perform an action such as posting a message. Yesterday, we noticed an issue where, after restart, the `guildMemberUpdate` event was not firing. I realized this was because we weren't using partials: https://github.com/discordjs/discord.js/issues/5685

After adding in the `GUILD_MEMBER` partial, however, I am struggling to determine new roles that were added to a member. Specifically, when `guildMemberUpdate` is fired with a partial `oldMember` value, it doesn't include any role information for that member. Then, when I fetch the `oldMember` upon receiving the event, the fetched data contains all roles present on `newMember`. This makes it so that I can't tell which role was just added or removed. 

This presents an issue whenever my bot restarts and the first event it seems from a given member is a `guildMemberUpdate` due to a role change.

Do y'all have any workarounds or suggestions for this? Thanks in advance! 

### Code sample

```typescript
const client = new Client({
    [
      Intents.FLAGS.GUILDS,
      Intents.FLAGS.GUILD_MESSAGES,
      Intents.FLAGS.GUILD_MEMBERS,
      Intents.FLAGS.GUILD_MESSAGE_REACTIONS,
    ],
      restTimeOffset: 0,
      partials: ["GUILD_MEMBER"],
});
    
client.on("ready", async () => {
    console.log(
    `Logged in as bot!`
    );
});

client.on("guildMemberUpdate", async (oldMember, newMember) => {
    // NOTE: oldMember partial does not contain any roles 
    // if oldMember NOT a partial, it contains different roles than newMember
    if (oldMember.partial) {
        oldMember = await oldMember.fetch();
        // oldMember now contains same role info as newMember
    }
    
    await processRoleChange(
        newMember,
        oldMember.roles.cache,
        newMember.roles.cache,
        this.executor
    );
});
    
async function processRoleChange(
    member: GuildMember,
    oldRoles: Collection<string, Role>,
    newRoles: Collection<string, Role>,
    ) {
    // NOTE: THIS BREAKS BECAUSE ROLES ARE IDENTICAL
    const rolesAdded = newRoles.filter(
        (r) => !oldRoles.find((or) => or.id === r.id)
    );
    const rolesRemoved = oldRoles.filter(
        (r) => !newRoles.find((nr) => nr.id === r.id)
    );
    if (rolesAdded.size > 0 || rolesRemoved.size > 0) {
        console.log("do other stuff...")
    } // otherwise no role change, do nothing
}
```

### Package version

13.8.0

### Node.js version

16.13.1

### Operating system

MacOS 12.0.1

### Priority this issue should have

Medium (should be fixed soon)

### Which partials do you have configured?

GuildMember

### Which gateway intents are you subscribing to?

GuildMembers

**Comment by Jiralite** [comment]
> Specifically, when `guildMemberUpdate` is fired with a partial `oldMember` value, it doesn't include any role information for that member.

Seems you know what the issue is. It then follows that you should ensure all guild members in the server are cached by fetching them or utilise a database for which user ids are against role ids.

**Reply by pranavbadami** [reply]
Hey @Jiralite , thanks for the response! Hmm ok, gotcha, I was curious if there was a way to do this without fetching all the members. I will take a look at fetching members so that they're cached. My bot is still pretty small so this should hopefully suffice for now.
